src/Microsoft.Health.Core/Features/Health/AsyncData.cs

```diff
@@ -0,0 +1,38 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Core.Features.Health;
+
+/// <summary>
+/// Asynchronously waits for the data to be set before the data can be retrieved
+/// </summary>
+/// <typeparam name="T">The data type</typeparam>
+public class AsyncData<T> where T : class
+{
+    private volatile T _cachedData;
+    private readonly TaskCompletionSource _init = new TaskCompletionSource();
+
+    public async Task<T> GetCachedData(CancellationToken cancellationToken = default)
+    {
+        await _init.Task.WaitAsync(cancellationToken).ConfigureAwait(false);
+        return _cachedData;
+    }
+
+    public void SetCachedData(T cachedData)
+    {
+        EnsureArg.IsNotNull(cachedData, nameof(cachedData));
+
+        _cachedData = cachedData;
+
+        if (!_init.Task.IsCompleted)
+        {
+            _init.TrySetResult();
```

Comment:
I am unsure about the thread-safety implications of this class's method. It may be best to try to use a `lock`

Reply:
I don't know if that is really necessary. It looks like TaskCompletionSource promises thread safety (https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource-1?view=net-7.0#thread-safety), and if I look under the hood at `TrySetResult` it does seem like its handling it for us.
(https://github.com/microsoft/referencesource/blob/master/mscorlib/system/threading/Tasks/TaskCompletionSource.cs#L296)
```
        public bool TrySetResult(TResult result)
        {
            bool rval = m_task.TrySetResult(result);
            if (!rval && !m_task.IsCompleted) SpinUntilCompleted();
            return rval;
        }
```
Where `SpinUntilCompleted` is there to wait in case another thread is trying to set the result

Reply:
Ah, I missed this remark in the docs! Thanks for checking!